File: Labs/Lab9.cpp
#include <stack>
#include <catch2/catch_test_macros.hpp>
#include <queue>

//1
class MyQueue {
    std::stack<int> in_stack, out_stack;

public:
    void push(int x) {
        in_stack.push(x);
    }

    int pop() {
        if (out_stack.empty()) {
            while (!in_stack.empty()) {
                out_stack.push(in_stack.top());
                in_stack.pop();
            }
        }
        int top = out_stack.top();
        out_stack.pop();
        return top;
    }

    int peek() {
        if (out_stack.empty()) {
            while (!in_stack.empty()) {
                out_stack.push(in_stack.top());
                in_stack.pop();
            }
        }
        return out_stack.top();
    }

    bool empty() {
        return in_stack.empty() && out_stack.empty();
    }
};

TEST_CASE("Queue Using Stacks") {
    MyQueue queue;
    queue.push(1);
    queue.push(2);
    REQUIRE(queue.peek() == 1);
    REQUIRE(queue.pop() == 1);
    REQUIRE(!queue.empty());
}


//2
class MyStack {
    std::queue<int> queue;

public:
    void push(int x) {
        queue.push(x);
        for (int i = 0; i < queue.size() - 1; ++i) {
            queue.push(queue.front());
            queue.pop();
        }
    }

    int pop() {
        int top = queue.front();
        queue.pop();
        return top;
    }

    int top() {
        return queue.front();
    }

    bool empty() {
        return queue.empty();
    }
};

TEST_CASE("Stack Using Queues") {
    MyStack stack;
    stack.push(1);
    stack.push(2);
    REQUIRE(stack.top() == 2);
    REQUIRE(stack.pop() == 2);
    REQUIRE(!stack.empty());
}

//3
#include <unordered_set>
#include <catch2/catch_test_macros.hpp>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

bool hasCycle(ListNode* head) {
    ListNode* slow = head;
    ListNode* fast = head;

    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;

        if (slow == fast) {
            return true;
        }
    }

    return false;
}

TEST_CASE("Detect Linked List Cycle") {
    ListNode* head = new ListNode(3);
    ListNode* node2 = new ListNode(2);
    ListNode* node0 = new ListNode(0);
    ListNode* node4 = new ListNode(-4);
    head->next = node2;
    node2->next = node0;
    node0->next = node4;
    node4->next = node2;  // Creates a cycle

    REQUIRE(hasCycle(head) == true);

    ListNode* no_cycle_head = new ListNode(1);
    no_cycle_head->next = new ListNode(2);
    REQUIRE(hasCycle(no_cycle_head) == false);
}

File: Labs/Lab9.py
import unittest
from collections import deque

#1
class MyQueue:
    def __init__(self):
        self.in_stack = []
        self.out_stack = []
    
    def push(self, x: int):
        self.in_stack.append(x)
    
    def pop(self) -> int:
        if not self.out_stack:
            while self.in_stack:
                self.out_stack.append(self.in_stack.pop())
        return self.out_stack.pop()
    
    def peek(self) -> int:
        if not self.out_stack:
            while self.in_stack:
                self.out_stack.append(self.in_stack.pop())
        return self.out_stack[-1]
    
    def empty(self) -> bool:
        return not self.in_stack and not self.out_stack

# Unit Test
class TestMyQueue(unittest.TestCase):
    def test_queue(self):
        queue = MyQueue()
        queue.push(1)
        queue.push(2)
        self.assertEqual(queue.peek(), 1)
        self.assertEqual(queue.pop(), 1)
        self.assertFalse(queue.empty())

if __name__ == "__main__":
    unittest.main()


#2
class MyStack:
    def __init__(self):
        self.queue = []

    def push(self, x: int):
        self.queue.append(x)
        for _ in range(len(self.queue) - 1):
            self.queue.append(self.queue.pop(0))

    def pop(self) -> int:
        return self.queue.pop(0)

    def top(self) -> int:
        return self.queue[0]

    def empty(self) -> bool:
        return not self.queue

# Unit Test
class TestMyStack(unittest.TestCase):
    def test_stack(self):
        stack = MyStack()
        stack.push(1)
        stack.push(2)
        self.assertEqual(stack.top(), 2)
        self.assertEqual(stack.pop(), 2)
        self.assertFalse(stack.empty())

if __name__ == "__main__":
    unittest.main()

#3
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

def hasCycle(head: ListNode) -> bool:
    slow, fast = head, head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
    return False

# Unit Test
class TestHasCycle(unittest.TestCase):
    def test_cycle(self):
        head = ListNode(3)
        node2 = ListNode(2)
        node0 = ListNode(0)
        node4 = ListNode(-4)
        head.next = node2
        node2.next = node0
        node0.next = node4
        node4.next = node2  # Creates a cycle
        self.assertTrue(hasCycle(head))

        no_cycle_head = ListNode(1)
        no_cycle_head.next = ListNode(2)
        self.assertFalse(hasCycle(no_cycle_head))

if __name__ == "__main__":
    unittest.main()